domain/application/state/unit.go

```diff
@@ -1084,161 +1083,525 @@ func (st *State) UpdateCAASUnit(ctx context.Context, unitName coreunit.Name, par
+				}
+
+				constraintSpace := setConstraintSpace{ConstraintUUID: cUUIDStr, Space: space.SpaceName, Exclude: space.Exclude}
+				if err := tx.Query(ctx, insertConstraintSpacesStmt, constraintSpace).Run(); err != nil {
+					return errors.Capture(err)
+				}
+			}
+		}
+
+		if cons.Zones != nil {
+			for _, zone := range *cons.Zones {
+				constraintZone := setConstraintZone{ConstraintUUID: cUUIDStr, Zone: zone}
+				if err := tx.Query(ctx, insertConstraintZonesStmt, constraintZone).Run(); err != nil {
+					return errors.Capture(err)
+				}
+			}
+		}
+
+		return errors.Capture(
+			tx.Query(ctx, insertUnitConstraintsStmt, setUnitConstraint{
+				UnitUUID:       inUnitUUID.String(),
+				ConstraintUUID: cUUIDStr,
+			}).Run(),
+		)
+	})
+}
+
+// SetUnitPresence marks the presence of the specified unit, returning an error
+// satisfying [applicationerrors.UnitNotFound] if the unit doesn't exist.
+// The unit life is not considered when making this query.
+func (st *State) SetUnitPresence(ctx context.Context, name coreunit.Name) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	unit := unitName{Name: name}
+	var uuid unitUUID
+
+	queryUnit := `SELECT &unitUUID.uuid FROM unit WHERE name = $unitName.name;`
+	queryUnitStmt, err := st.Prepare(queryUnit, unit, uuid)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	recordUnit := `
+INSERT INTO unit_agent_presence (*) VALUES ($unitPresence.*)
+ON CONFLICT(unit_uuid) DO UPDATE SET
+	last_seen = excluded.last_seen;
+`
+	presence := unitPresence{
+		LastSeen: st.clock.Now(),
+	}
+	recordUnitStmt, err := st.Prepare(recordUnit, presence)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		if err := tx.Query(ctx, queryUnitStmt, unit).Get(&uuid); errors.Is(err, sql.ErrNoRows) {
+			return applicationerrors.UnitNotFound
+		} else if err != nil {
+			return errors.Capture(err)
+		}
+
+		presence.UnitUUID = uuid.UnitUUID
+
+		if err := tx.Query(ctx, recordUnitStmt, presence).Run(); err != nil {
+			return errors.Capture(err)
+		}
+		return nil
+	})
+	if err != nil {
+		return errors.Capture(err)
+	}
+	return nil
+}
+
+// DeleteUnitPresence removes the presence of the specified unit. If the
+// unit isn't found it ignores the error.
+// The unit life is not considered when making this query.
+func (st *State) DeleteUnitPresence(ctx context.Context, name coreunit.Name) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	unit := unitName{Name: name}
+	var uuid unitUUID
+
+	queryUnit := `SELECT &unitUUID.uuid FROM unit WHERE name = $unitName.name;`
+	queryUnitStmt, err := st.Prepare(queryUnit, unit, uuid)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	var presence unitPresence
+	deleteStmt, err := st.Prepare(`
+DELETE FROM unit_agent_presence
+WHERE unit_uuid = $unitPresence.unit_uuid
+`, presence)
+	if err != nil {
+		return errors.Capture(err)
+	}
```

Comment:
This can be done in one query with a `JOIN` no?

Reply:
No, but you can do it using an inner select.